validator/client/src/test/java/tech/pegasys/teku/validator/client/DoppelgangerDetectorTest.java

```diff
@@ -0,0 +1,278 @@
+  private final GenesisDataProvider genesisDataProvider = mock(GenesisDataProvider.class);
+  private final ValidatorIndexProvider validatorIndexProvider = mock(ValidatorIndexProvider.class);
+  private final LogCaptor logCaptor = LogCaptor.forClass(DoppelgangerDetector.class);
+  private final String doppelgangerStartLog = "Starting doppelganger detection...";
+  private final String doppelgangerTimeoutLog =
+      "Validators Doppelganger Detection timeout reached. Some technical issues prevented the validators doppelganger detection from running correctly. Please check the logs and consider performing a new validators doppelganger check.";
+
+  private final Duration checkDelay = Duration.ofSeconds(2);
+  private final Duration timeout = Duration.ofMinutes(15);
+
+  private DoppelgangerDetector doppelgangerDetector;
+
+  @BeforeEach
+  public void setup() throws IOException {
+    when(genesisDataProvider.getGenesisTime()).thenReturn(SafeFuture.completedFuture(UInt64.ZERO));
+    when(validatorIndexProvider.getValidatorIndices())
+        .thenReturn(SafeFuture.completedFuture(IntArrayList.of(1, 2, 3)));
+    doppelgangerDetector =
+        new DoppelgangerDetector(
+            statusLog,
+            asyncRunner,
+            validatorApiChannel,
+            validatorIndexProvider,
+            spec,
+            timeProvider,
+            genesisDataProvider,
+            checkDelay,
+            timeout);
+  }
+
+  @Test
+  public void shouldNotDetectDoppelganger() {
+    when(validatorApiChannel.getValidatorsLiveness(any(), any()))
+        .thenReturn(
+            SafeFuture.completedFuture(
+                Optional.of(
+                    List.of(
+                        new ValidatorLivenessAtEpoch(
+                            UInt64.valueOf(4), dataStructureUtil.randomEpoch(), true),
+                        new ValidatorLivenessAtEpoch(
+                            UInt64.valueOf(5), dataStructureUtil.randomEpoch(), true)))));
+    SafeFuture<Map<Integer, BLSPublicKey>> doppelgangerDetectorFuture =
+        doppelgangerDetector.performDoppelgangerDetection();
+    asyncRunner.executeQueuedActions();
+    timeProvider.advanceTimeBySeconds(120);
+    asyncRunner.executeQueuedActions();
+    assertThat(logCaptor.getLogEvents().size()).isEqualTo(5);
+    expectLogMessage(logCaptor.getLogEvents().get(0), "INFO", doppelgangerStartLog);
+    expectLogMessage(logCaptor.getLogEvents().get(1), "INFO", doppelgangerDetectorStartEpochLog(0));
+    expectLogMessage(logCaptor.getLogEvents().get(2), "INFO", performingDoppelgangerCheckLog(0, 1));
+    expectLogMessage(
+        logCaptor.getLogEvents().get(3),
+        "INFO",
+        "No validators doppelganger detected for epoch 0, slot 1");
+    expectLogMessage(
+        logCaptor.getLogEvents().get(4),
+        "INFO",
+        "No validators doppelganger detected after 2 epochs. Stopping doppelganger detection.");
+    assertThat(doppelgangerDetectorFuture).isCompleted();
```

Comment:
you can remove those `isCompleted();` if later you check with `isCompletedWithValue`

Reply:
Done